# Week 06. Ensemble

과제는 총 3개의 cell과 markdown 하나를 채워주셔야합니다

1~3 문제는 영상과 실습자료에 나와있는 것들을 적절히 응용하시면 됩니다

마지막 4번의 경우 본인의 생각을 마크다운으로 작성하는 부분입니다

주석의 경우 이미지, 테이블 등의 표현이 어려운 관계로 받지 않겠습니다

### 실습 00. Load dataset

이번 과제에 사용할 dataset은 The census - Adult dataset입니다

ML dataset repository인 UCI에서 받을 수 있습니다

파일의 마지막 열에는 target 정보가 기록되어있습니다

데이터를 불러온 후 variables와 target으로 분리해주세요

In [1]:
import pandas as pd
import numpy as np
import os 

df=pd.read_csv('adult.csv')

target = df.iloc[:,-1]
variables= df.iloc[:,:-1]

주어진 데이터를 학습하여 보지않은 test data에 대해 잘 분류해야합니다

아래 train test split 함수를 사용하여 train set과 test set으로 분리해주세요

이후 모든 모델 학습의 결과는 test set으로 평가해주셔야 합니다

In [2]:
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(variables.value,target,test_size=0.2)

### 실습 01. Voting ensemble

Dataset을 약 200개의 base learner에 학습시킨 뒤, 정확도를 출력해주세요

실습한 3가지 모델을 전부 사용해주셔야 하지만, 각 모델의 hyper parameter는 자유롭게 사용해주세요

Hard voting과 soft voting을 각각 시행하여 결과를 출력해주셔야 합니다

In [3]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB

from sklearn.ensemble import VotingClassifier

models = [('GNBC', GaussianNB())]+[(f'kNN-{n}',KNeighborsClassifier(n)) for n in range(1,30,1)]+\
            [(f'DT-{_}',DecisionTreeClassifier()) for _ in range(170)]

copy_models = models.copy()

ensemble=VotingClassifier(models, n_jobs=-1, voting='hard')
ensemble.fit(x_train.values, y_train.values.ravel())

print(f" Hard ensemble score : {ensemble.score(x_test, y_test)*100: .5}%")

 Hard ensemble score :  82.096%


In [4]:
models = [('GNBC', GaussianNB())]+[(f'kNN-{n}',KNeighborsClassifier(n)) for n in range(1,30,1)]+\
            [(f'DT-{_}',DecisionTreeClassifier()) for _ in range(170)]

copy_models = models.copy()

ensemble1=VotingClassifier(models, n_jobs=-1, voting='soft')
ensemble1.fit(x_train.values, y_train.values.ravel())

print(f" Soft ensemble score : {ensemble1.score(x_test, y_test)*100: .5}%")

 Soft ensemble score :  81.994%


### 실습 02. Bootstrap Aggregating (bagging)

bagging을 실습해보겠습니다

Random forest는 bagging을 지원하지만 다양한 classifier를 사용하는 경우 bagging을 직접 구현해야합니다

bagging을 하기위한 package들도 존재하지만 numpy를 사용하여 bagging을 간단히 적용할 수 있습니다

먼저 sample만 무작위로 복원추출 하는 함수를 작성해주세요

np.random.choice 함수를 사용하여 slicing하면 간단히 구현할 수 있습니다

In [5]:
def sample_bootstrap(mat,bootstrap_ratio=0.6):
    data= [data for data in range(mat)]
    bootstrap_mat = np.random.choice(data, mat)
    return bootstrap_mat

In [1]:
## 조교 코드 
def sample_bootstrap(mat,bootstrap_ratio=0.6):
    row_idx=np.random.choice(range(data.shape[0]),
                            int(data.shape[0]*bootstrap_ratio),
                            False)
    bootstrap_x = data[row_idx]
    bootstrap_y = target.values[row_idx]
    return bootstrap_x, bootstrap_y

x_train.shape, sample_bootstrap(x_train, y_train)

NameError: name 'x_train' is not defined

이어서 sample과 함께 variables들도 복원추출할 수 있는 함수를 작성해주세요

위와 같이 함수를 사용하여 row뿐만 아니라 column도 slicing하는 것으로 구현할 수 있습니다

In [15]:
## 조교 코드 
def sample_bootstrap(mat,bootstrap_ratio=0.9):
    col_idx=np.random.choice(range(data.shape[1]),
                            int(data.shape[1]*bootstrap_ratio),
                            False)
    row_idx=np.random.choice(range(data.shape[0]),
                            int(data.shape[0]*bootstrap_ratio),
                            False)
    bag_idx.append([row_idx, col_idx])
    
    bootstrap_x = data[row_idx][:,col_idx]
    bootstrap_y = target.values[row_idx]
    return bootstrap_x, bootstrap_y

x_train.shape.sample_bootstrap(x_train, y_train)

In [ ]:
bag_idx=[]  # 이거 없으면 에러남 

In [ ]:
x_train.shape, bootstrap(x_train, y_train), len(bag_idx)

위에서 작성한 두 bootstrap 방법을 사용하여 실습 1에서 만든 모델들을 학습시키고, 적절하게 aggregating하여 결과를 도출해주세요

### sample bagging

In [ ]:
models = [('GNBC', GaussianNB())]+[(f'kNN-{n}',KNeighborsClassifier(n)) for n in range(1,30,1)]+\
            [(f'DT-{_}',DecisionTreeClassifier()) for _ in range(170)]

preds_list=[]
for clf in models:
    preds_list.append(clf[1].predict(x_test))
    
((np.array(preds_list).sum(0)>100)==y_test).mean()*100

### sample & feature bagging

In [ ]:
bag_idx=[]
models = [('GNBC', GaussianNB())]+[(f'kNN-{n}',KNeighborsClassifier(n)) for n in range(1,30,1)]+\
            [(f'DT-{_}',DecisionTreeClassifier()) for _ in range(170)]

In [ ]:
preds_list = []
for clf,(row, col) in zip(models, bag_idx):
    if clf[0] == 'GNBC':
        preds_list.append(clf[1].predict(x_test))
    else:
        preds_list.append(clf[1].predict(x_test[:,col]))
((np.array(preds_list).sum(0)>100)==y_test).mean()*100

정보의 손실이 나서 오히려 정확도가 떨어진다.


우리는 지난 실슴들을 통해 다양한 preprocessing을 실습하였습니다

앞서 생성한 모델을 학습시킬때 성능을 개선시킬 방법이 있는지 고민해보시고 방법론을 아래 markdown으로 작성해주세요

kNN을 위해서 standard scle을 하면 좀더 좋은 성능을 얻을수 있다.

### 실습 03. Random forest & gradient boosting machine

Random forest와 gradient boosting machine에 대해 사용해보시고 결과를 출력해주세요

hyper parameter는 자유롭게 사용해주시면 됩니다

그리고 feature importance 구하여 각 base,weak learner들이 어떻게 instance를 분리하였는지 간단히 설명해주세요

In [8]:
from sklearn.ensemble import RandomForestClassifier,GradientBoostingClassifier
import plotly.graph_objects as go

rf=RandomForestClassifier(200,n_jobs=-1)
rf.fit(x_train, y_train)

gbc = GradientBoostingClassifier(n_estimators=100000,n_iter_no_change=1000, varbose =1)
gbc.fit(x_train,y_train)

GradientBoostingClassifier(n_estimators=100000, n_iter_no_change=1000)

In [ ]:
(rf.predict(x_test)==y_test).mean()*100

In [ ]:
(gbc.predict(x_test)==y_test).mean()*100

In [9]:
rf_importances = rf.feature_importances_

gb_importances = gbc.feature_importances_

indices = np.argsort(gb_importances)[::-1]  

cols = df.columns[indices]
cols = ['feature ' + i for i in cols]

fig = go.Figure(data=[
    
    go.Bar(name='Gradient Boosting', x=cols,
          y=gb_importances[indices],
          ),
    go.Bar(name='Random Forest', x=cols,
          y=rf_importances[indices],
          ),
])

fig_options={'layout':dict(template='simple_white',
                          width=800, height=500,
                          xaxis_title='Feature names',
                          yaxis_title = 'importances',
                          barmode='group')}
fig.update(fig_options)

fig.show()

Random Forest는 0번과 2번의 importance가 높고 8,9,13번의 importance가 낮게 측정되었다.  
GB는 4번, 5번, 10번의 importance가 높고  1,3,8,9,13번의 importance가 낮게 측정되었다.  
instance를 분리할때 Random forest는 0번과 2번 Feature 위주로 분리를 하고 GB는 4번,5번,10번 Feature 위주로 분리한다고 볼수 있다.  

### 실습 04. Stacking

우리는 다양한 base learner를 사용하여 ensemble을 시도하였습니다

그리고 Ensemble의 결과를 얻기 위해 majority voting, argmax $p(\hat{y})$ 등을 시도하였습니다

다른 측면에서 voting 자체도 기계학습으로 대체하는 기법이 있습니다

이것을 stacking이라고 부릅니다

간단한 모델들을 사용하여 base learner들의 예측 확률 값을 최종 모델의 input으로 사용하는 stacking model을 생성해주세요

Stacking model은 기존과 같은 target을 학습하며, 이렇게 학습된 모델이 어떤 예측 성능을 보이는지 보여주세요

In [50]:
DT = DecisionTreeClassifier(criterion='entropy')
rf = RandomForestClassifier()
kNN = KNeighborsClassifier(5)

Stacking = DecisionTreeClassifier(criterion='gini')

DT.fit(x_train, y_train)
rf.fit(x_train, y_train)
kNN.fit(x_train, y_train)

DT_pred = DT.predict(x_test)
rf_pred = rf.predict(x_test)
kNN_pred = kNN.predict(x_test)



In [56]:
new_data = np.array([DT_pred, rf_pred, kNN_pred])
new_data = np.transpose(new_data)
Stacking.fit(new_data, y_test)
print("score : {:.3f}".format(Stacking.score(new_data,y_test)*100))

score : 86.027


Decision Tree(entropy)와 RandomForest, kNN을 base learner로 하고 Stacking model이 Decision Tree(gini)로 만들어 학습시켰다. 
실습 01에서 만든  Hard ensemble모델과 Soft ensemble 모델의 score는 각각 82.096%, 81.994%인 반면에  
Stacking model은 86.027%로 가장 높은 예측성능을 보여준다. 
